drivers: serial: Infineon PDL based UART Async. driver

ChanChihHsiang opened the pull request:
Adding Infineon PDL based driver to support UART Async. on the PSE84 and PSC3. 
Add support for the following apps.:
samples\drivers\uart\async_api
tests/drivers/uart/uart_async_api

chanchihhsia committed:
Drivers: Serial: Add PDL based UART Async. for PSE84 and PSC3

- add UART Async. driver code
- Remove references to cat1-uart-pdl from devicetree references

Signed-off-by: Kevin Chan <kevin.chan3@infineon.com>

chanchihhsia committed:
tests: drivers: uart: uart_async_api: support pse84 and psc3

- add overlay and .conf files in order to run uart_async_api app.

Signed-off-by: Kevin Chan <kevin.chan3@infineon.com>

chanchihhsia committed:
samples: drivers: uart: async_api: support PSC3 and PSE84

- add overlay files in order to run async_api app.

Signed-off-by: Kevin Chan <kevin.chan3@infineon.com>

JarmouniA reviewed (dismissed):
Please use this chance to remove the binding
https://github.com/zephyrproject-rtos/zephyr/blob/613d228002ac134ab432d3d4f9280ef58c00bec4/dts/bindings/serial/infineon%2Ccat1-uart-pdl.yaml
and just use the compatible `infineon,cat1-uart` will all drivers for this device. See https://github.com/zephyrproject-rtos/zephyr/pull/98035#issuecomment-3472246866 for context

jsbatch commented on samples/drivers/uart/async_api/boards/kit_pse84_eval_pse846gps2dbzc4a_m33_ns.overlay:
M33 NS mode is not currently supported in upstream Zephyr repository.  Please remove these overlays.

ChanChihHsiang replied:
removed

chanchihhsia committed:
drivers: dma: add UART async. logic

- Modified DMA configuration and logic for UART Async.

Signed-off-by: Kevin Chan <kevin.chan3@infineon.com>

jsbatch committed:
boards: Infineon: Removing dts references to infineon,cat1-uart-pdl

Removing references to infineon,cat1-uart-pdl from device tree board
files.  The differentiation between PDL and UART based drivers is done
by a Kconfig option.  See discussion in this PR for context:

https://github.com/zephyrproject-rtos/zephyr/pull/98035

Signed-off-by: John Batch <john.batch@infineon.com>

jsbatch commented:
I have pushed changes to address the comment from @JarmouniA because I have been working in parallel on these changes for several drivers.  See the linked PR conversation for full context.

ChanChihHsiang commented on drivers/dma/dma_ifx_cat1_pdl.c:
```diff
@@ -366,31 +379,114 @@ int ifx_cat1_dma_reload(const struct device *dev, uint32_t channel, uint32_t src
 	const struct ifx_cat1_dma_config *const cfg = dev->config;
 
-	if (channel >= cfg->num_channels) {
+	if (channel >= DMA_CH_NUM) {
```
Corrected

JarmouniA reviewed (commented):
Commit message sign-off should only contain your name and email.

sreeramIfx commented:
Please also fix the commit mesage of the first commit

ChanChihHsiang commented:
> Please also fix the commit mesage of the first commit

Removed last line. Thanks

ChanChihHsiang commented on drivers/dma/dma_ifx_cat1_pdl.c:
```diff
@@ -366,6 +379,72 @@ int ifx_cat1_dma_reload(const struct device *dev, uint32_t channel, uint32_t src
+	while ((next_descr != NULL) && (next_descr != curr_descr)) {
+		x_size = Cy_DMA_Descriptor_GetXloopDataCount(next_descr);
+		y_size = Cy_DMA_Descriptor_GetYloopDataCount(next_descr);
+		transferred_data_size += (y_size != 0U) ? x_size * y_size : x_size;
```
Corrected

sreeramIfx commented on drivers/serial/uart_ifx_cat1_pdl.c:
```diff
@@ -34,6 +34,39 @@ LOG_MODULE_REGISTER(uart_ifx_cat1, CONFIG_UART_LOG_LEVEL);
+#include <cy_trigmux.h>
+
+struct ifx_cat1_dma_stream {
+	const struct device *dev_dma;
```
```suggestion
	const struct device *dma_dev;
```
to be consistent with the below two variables

ChanChihHsiang replied:
corrected

sreeramIfx commented on drivers/serial/uart_ifx_cat1_pdl.c:
```diff
@@ -641,6 +677,467 @@ static const cy_stc_scb_uart_config_t _uart_default_config = {
+	const struct device *dev_dma = data->async.dma_tx.dev_dma;
+	int err;
+
+	unsigned int key = irq_lock();
```
Can we move irq_lock further down after the error checks?

sreeramIfx commented on drivers/serial/uart_ifx_cat1_pdl.c:
```diff
@@ -641,6 +677,467 @@ static const cy_stc_scb_uart_config_t _uart_default_config = {
+		event.data.rx_buf.buf = data->async.dma_rx.buf;
+		data->async.dma_rx.buf = NULL;
+		data->async.dma_rx.buf_len = 0;
+	}
```
```suggestion
	event.data.rx_buf.buf = (buffer_type == NEXT_BUFFER) ? data->async.rx_next_buf:data->async.dma_rx.buf;
	data->async.dma_rx.buf = NULL;
	data->async.dma_rx.buf_len = 0;
```

sreeramIfx commented on drivers/serial/uart_ifx_cat1_pdl.c:
```diff
@@ -641,6 +677,467 @@ static const cy_stc_scb_uart_config_t _uart_default_config = {
+			data->async.dma_rx.counter = rx_rcv_len;
+			async_evt_rx_rdy(data);
+		}
+	}
```
```suggestion
	}
	
```

ChanChihHsiang replied:
corrected

sreeramIfx commented on drivers/serial/uart_ifx_cat1_pdl.c:
```diff
@@ -641,6 +677,467 @@ static const cy_stc_scb_uart_config_t _uart_default_config = {
+
+	data->async.cb = callback;
+	data->async.user_data = user_data;
+	data->async.dma_tx.dma_cfg.user_data = (void *)dev;
```
Why are we setting the dma_tx userdata only?

ChanChihHsiang commented:
@teburd Hi can you review this PR

teburd commented on boards/infineon/kit_pse84_eval/kit_pse84_eval_common.dtsi:
```diff
@@ -45,7 +45,7 @@
 
 uart2: &scb2 {
-	compatible = "infineon,cat1-uart-pdl";
+	compatible = "infineon,cat1-uart";
```
Lets drop the cat1 name

Non-blocking
